test/dispatch.jl

```diff
@@ -0,0 +1,180 @@
+
+    @testset "ambiguous" begin
+        @test !type_morespecific(Tuple{Int,AbstractFloat}, Tuple{Integer,Float64})
+        @test !type_morespecific(Tuple{Integer,Float64}, Tuple{Int,AbstractFloat})
+    end
+
+    @testset "length" begin
+        @test type_morespecific(Tuple{Int,Int}, Tuple{Int,Vararg{Int}})
+    end
+
+    @testset "union" begin
+        @test type_morespecific(Tuple{Int}, Tuple{Union{Int,Float64}})
+        @test type_morespecific(Tuple{Union{Int,Float64}}, Tuple{Integer})
+    end
+
+    @testset "union all" begin
+        @test type_morespecific(Tuple{Int}, Tuple{T} where T <: Integer)
+
+        @test !type_morespecific(Tuple{Integer}, Tuple{T} where T <: Integer)
+        @test !type_morespecific(Tuple{T} where T <: Integer, Tuple{Integer})
+
+        @test type_morespecific(Tuple{Type{Integer}}, Tuple{Type{T}} where T <: Integer)
+        @test !type_morespecific(Tuple{Type{T}} where T <: Integer, Tuple{Type{Integer}})
+    end
+end
+
+@testset "anonymous_signature" begin
+    @testset "diagonal dispatch" begin
+        diag(::T, ::T) where T <: Integer = nothing
+
+        diag_method = first(methods(diag, (Int, Int)))
+        diag_sig = Tuple{typeof(diag),T,T} where T <: Integer
+        @test diag_method.sig == diag_sig
+
+        @test anonymous_signature(Tuple{typeof(diag),Int,Int}) == Tuple{Int,Int}
+        @test anonymous_signature(diag_sig) == Tuple{T,T} where T <: Integer
+        @test anonymous_signature(diag_method) == Tuple{T,T} where T <:Integer
+    end
+
+    @testset "triangular dispatch" begin
+        tri(::Vector{T}, ::S) where {T, S<:T} = nothing
+
+        tri_method = first(methods(tri, (Vector, Int)))
+        tri_sig = Tuple{typeof(tri),Vector{T},S} where {T, S<:T}
+        @test tri_method.sig == tri_sig
+
+        @test anonymous_signature(Tuple{typeof(tri),Vector,Int}) == Tuple{Vector,Int}
+        @test anonymous_signature(tri_sig) == Tuple{Vector{T},S} where {T, S<:T}
+        @test anonymous_signature(tri_method) == Tuple{Vector{T},S} where {T, S<:T}
+    end
+end
+
+@testset "anon_morespecific" begin
+    plus_method = first(methods(+, (Int, Int)))
+    max_method = first(methods(max, (Int, Int)))
+
+    @test !type_morespecific(plus_method.sig, max_method.sig)
+    @test anon_morespecific(plus_method, max_method)
+end
+
+@testset "dispatch" begin
+    @testset "no functions" begin
+        m, f = dispatch(Function[])
+        @test f === nothing
+        @test m === nothing
+    end
+
+    @testset "more specific function" begin
+        funcs = [
+            (::Int) -> 1
+            (::Integer) -> 2
+        ]
+
+        m, f = dispatch(funcs, zero(Int))
+        @test f == funcs[1]
+        @test m == first(methods(funcs[1]))
+
+
+        m, f = dispatch(funcs, zero(UInt))
+        @test f == funcs[2]
+        @test m == first(methods(funcs[2]))
+    end
+
+    @testset "more specific method" begin
+        # Note: `methods` returns methods in order of most specific to least.
```

Comment:
I don't think this is a promise, it is not in the docstring, 
and similar things have changed before https://github.com/JuliaLang/julia/pull/32776

Further in ambigious cases it is not controlled:

```
g(::Irrational) = 3
g(::Signed) = 4
g(::Union{typeof(pi), Integer}) =  1
g(::Union{AbstractIrrational, Int}) =  2


g2(::Union{typeof(pi), Integer}) =  1
g2(::Union{AbstractIrrational, Int}) =  2
g2(::Signed) = 4
g2(::Irrational) = 3
```

```

julia> methods(g)
# 4 methods for generic function "g":
[1] g(::Irrational) in Main at REPL[1]:1
[2] g(::Union{Int64, AbstractIrrational}) in Main at REPL[4]:1
[3] g(::Signed) in Main at REPL[2]:1
[4] g(::Union{Irrational{:π}, Integer}) in Main at REPL[3]:1

julia> methods(g2)
# 4 methods for generic function "g2":
[1] g2(::Signed) in Main at REPL[7]:1
[2] g2(::Union{Irrational{:π}, Integer}) in Main at REPL[5]:1
[3] g2(::Irrational) in Main at REPL[8]:1
[4] g2(::Union{Int64, AbstractIrrational}) in Main at REPL[6]:1
```

So I would rather not rely on it, and be somewhat more explict about what method we are getting